02-peer-protocol.md

```diff
@@ -1342,6 +1359,47 @@ it's simplest to only allow it to set fee levels; however, as the same
 fee rate applies to HTLC transactions, the receiving node must also
 care about the reasonableness of the fee.
 
+### Signing alternative feerates
+
+When a node sends `commitment_signed`, it may also sign alternative versions
+of the commitment using different feerates.
+
+1. type: 137 (`alternative_commitment_signed`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u32`:`feerate_per_kw`]
+   * [`signature`:`signature`]
+   * [`u16`:`num_htlcs`]
+   * [`num_htlcs*signature`:`htlc_signature`]
+
+#### Requirements
+
+A sending node:
+  - If `option_alternative_feerates` was negotiated:
+    - MUST send one `alternative_commitment_signed` for each feerate listed
+      in the corresponding `commitment_signed.alternative_feerates`.
```

Comment:
Note that HTLC-timeout transactions are CLTV locked (with independent CLTVs between htlcs). A CLTV might be hundreds of blocks into the future. Broadcasting a commitment tx at `alternative_feerate1` locks in that same feerate for all HTLCs too. (because the sighash in the htlc tx's input commits to the txid of the parent commit tx)

Unless we make it quadratic and pre-sign htlc txs at each `alternative_feerate` for commit txs at each `alternative_feerate`. That would mean creating `num_htlcs * len(alternative_feerates)**2` signatures at every committed channel state update.

Reply:
This is a very good remark. I went back and forth on that, the other option I had in mind that doesn't have this issue is to only sign alternative HTLC txs and not the commit tx. We keep a single commit tx (the one signed in `commitment_signed`, which will likely need some CPFP using the anchor) and sign alternative HTLC txs for each pending HTLC (and all these alternatives would spend the same output of the commit tx).

That means we need to handle the case where an HTLC tx doesn't make sense to be signed at some feerates (because its value goes entirely to fees): there are different ways of doing that (either skip them entirely or provide a dummy signature like `0x00...00`), it's not necessarily hard but is a bit messy.